api/api/src/main/scala/uk/ac/wellcome/platform/api/QueryParams.scala

```diff
@@ -0,0 +1,238 @@
+  // available here:
+  // https://doc.akka.io/docs/akka-http/current/routing-dsl/directives/custom-directives.html
+  def parse =
+    parameter(
+      (
+        "page".as[Int].?,
+        "pageSize".as[Int].?,
+        "workType".as[WorkTypeFilter] ?,
+        "items.locations.locationType".as[ItemLocationTypeFilter].?,
+        "production.dates.from".as[LocalDate].?,
+        "production.dates.to".as[LocalDate].?,
+        "language".as[LanguageFilter].?,
+        "genres.label".as[GenreFilter].?,
+        "subjects.label".as[SubjectFilter].?,
+        "include".as[V2WorksIncludes].?,
+        "aggregations".as[List[AggregationRequest]].?,
+        "sort".as[List[SortRequest]].?,
+        "sortOrder".as[SortingOrder].?,
+        "query".as[String].?,
+        "_queryType".as[String].?,
+        "_index".as[String].?,
+      )
+    ).tflatMap { args =>
+      val params = (MultipleWorksParams.apply _).tupled(args)
+      params.validationErrors match {
+        case Nil => provide(params)
+        case errs =>
+          reject(ValidationRejection(errs.mkString(", ")))
+            .toDirective[Tuple1[MultipleWorksParams]]
+      }
+    }
+
+  implicit val workTypeFilter: Decoder[WorkTypeFilter] =
+    decodeCommaSeperated.emap(strs => Right(WorkTypeFilter(strs)))
+
+  implicit val itemLocationTypeFilter: Decoder[ItemLocationTypeFilter] =
+    decodeCommaSeperated.emap(strs => Right(ItemLocationTypeFilter(strs)))
+
+  implicit val languageFilter: Decoder[LanguageFilter] =
+    decodeCommaSeperated.emap(strs => Right(LanguageFilter(strs)))
+
+  implicit val genreFilter: Decoder[GenreFilter] =
+    Decoder.decodeString.emap(str => Right(GenreFilter(str)))
+
+  implicit val subjectFilter: Decoder[SubjectFilter] =
+    Decoder.decodeString.emap(str => Right(SubjectFilter(str)))
+
+  implicit val aggregationsDecoder: Decoder[List[AggregationRequest]] =
+    decodeOneOfCommaSeperated(
+      "workType" -> AggregationRequest.WorkType,
+      "genres" -> AggregationRequest.Genre,
+      "production.dates" -> AggregationRequest.ProductionDate,
+      "subjects" -> AggregationRequest.Subject,
+      "language" -> AggregationRequest.Language,
+    )
+
+  implicit val sortDecoder: Decoder[List[SortRequest]] =
+    decodeOneOfCommaSeperated(
+      "production.dates" -> ProductionDateSortRequest
+    )
+
+  implicit val sortOrderDecoder: Decoder[SortingOrder] =
+    decodeOneOf(
+      "asc" -> SortingOrder.Ascending,
+      "desc" -> SortingOrder.Descending,
+    )
+}
+
+trait QueryParamsUtils extends Directives with TimeInstances {
+
+  implicit def unmarshaller[T](
+    implicit decoder: Decoder[T]): Unmarshaller[String, T] =
+    Unmarshaller.strict[String, T] { str =>
+      decoder.decodeJson(Json.fromString(str)) match {
+        case Left(err)    => throw new IllegalArgumentException(err.message)
+        case Right(value) => value
+      }
+    }
+
+  implicit val includesDecoder: Decoder[V2WorksIncludes] =
```

Comment:
Looking through this - we seems to have a few different ways of doing things for the same thing.
e.g. Aggregations are just a string enum to case Object enum map, which can be a list. So is Includes.

With includes we seem to have encoded the knowledge of the valid string values into that object. I think I prefer the knowledge to live here with the routing, removing any type-unsafe string checking out of the code. Could we do that with includes?

Reply:
@jamesgorrie good point that the query params file should have knowledge about decoding the strings. Does [this commit](https://github.com/wellcometrust/catalogue/pull/192/commits/553da588ae94617c77a920d9e120a1faa7f855fa) work for you